core/modules/nat.cc

```diff
@@ -18,173 +19,199 @@ using IpProto = bess::utils::Ipv4::Proto;
 using bess::utils::Udp;
 using bess::utils::Tcp;
 using bess::utils::Icmp;
-using bess::utils::CalculateChecksumIncremental16;
-using bess::utils::CalculateChecksumUnfoldedIncremental32;
-using bess::utils::CalculateChecksumUnfoldedIncremental16;
-using bess::utils::FoldChecksumIncremental;
+using bess::utils::ChecksumIncrement16;
+using bess::utils::ChecksumIncrement32;
+using bess::utils::UpdateChecksumWithIncrement;
+using bess::utils::UpdateChecksum16;
 
-const Commands NAT::cmds = {
-    {"add", "NATArg", MODULE_CMD_FUNC(&NAT::CommandAdd), 0},
-    {"clear", "EmptyArg", MODULE_CMD_FUNC(&NAT::CommandClear), 0}};
+CommandResponse NAT::Init(const bess::pb::NATArg &arg) {
+  for (const std::string &ext_addr : arg.ext_addrs()) {
+    be32_t addr;
+    bool ret = bess::utils::ParseIpv4Address(ext_addr, &addr);
+    if (!ret) {
+      return CommandFailure(EINVAL, "invalid IP address %s", ext_addr.c_str());
+    }
 
-inline Flow Flow::ReverseFlow() const {
-  if (proto == IpProto::kIcmp) {
-    return Flow(dst_ip, src_ip, icmp_ident, be16_t(0), IpProto::kIcmp);
-  } else {
-    return Flow(dst_ip, src_ip, dst_port, src_port, proto);
+    ext_addrs_.push_back(addr);
   }
-}
 
-std::string Flow::ToString() const {
-  return bess::utils::Format("%d %08x:%hu -> %08x:%hu", proto, src_ip.value(),
-                             src_port.value(), dst_ip.value(),
-                             dst_port.value());
-}
+  if (ext_addrs_.empty()) {
+    return CommandFailure(EINVAL,
+                          "at least one external IP address must be specified");
+  }
 
-CommandResponse NAT::Init(const bess::pb::NATArg &arg) {
-  InitRules(arg);
   return CommandSuccess();
 }
 
-CommandResponse NAT::CommandAdd(const bess::pb::NATArg &arg) {
-  InitRules(arg);
-  return CommandResponse();
-}
+static inline std::pair<bool, Endpoint> ExtractEndpoint(const Ipv4 *ip,
+                                                        const void *l4,
+                                                        NAT::Direction dir) {
+  IpProto proto = static_cast<IpProto>(ip->protocol);
 
-CommandResponse NAT::CommandClear(const bess::pb::EmptyArg &) {
-  rules_.clear();
-  flow_hash_.Clear();
+  if (likely(proto == IpProto::kTcp || proto == IpProto::kUdp)) {
+    // UDP and TCP share the same layout for port numbers
+    const Udp *udp = static_cast<const Udp *>(l4);
+    Endpoint ret;
 
-  return CommandResponse();
-}
+    if (dir == NAT::kForward) {
```

Comment:
The if statement will disappear at compile time. The functions take `dir` as a template parameter, so that the compiler will generate separate binary code for each forward and reverse direction. Eventually all the functions are inlined as two `ProcessBatch<kForward>()` and `ProcessBatch<kReverse>()`.